### ISSUE TITLE
Allow `bmp` & `gif` avatar uploads on Nav Apps

### ISSUE THREAD

**Issue by Beamanator** [Expensify/App#13038]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:
Break down in numbered steps

## Expected Result:
Describe what you think should've happened

## Actual Result:
Describe what actually happened

## Workaround:
Can the user still use Expensify without this being fixed? Have you informed them of the workaround?

## Platform:

Where is this issue occurring?

- Web
- iOS
- Android
- Desktop App
- Mobile Web

**Version Number:** 
**Reproducible in staging?:** 
**Reproducible in production?:** 
**Email or phone of affected tester (no customers):** 
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** Any additional supporting documentation
**Expensify/Expensify Issue URL:** 
**Issue reported by:** 
**Slack conversation:** 

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

**Comment by Beamanator** [comment]
While this issue currently exists, I would prefer waiting till https://github.com/Expensify/App/pull/12549 gets merged before working on this (most likely can be external) since in that PR we're enabling other image formats to be uploaded

**Comment by Beamanator** [comment]
https://github.com/Expensify/App/pull/12549 just got merged! I thinkkkkk we should still wait for it to get to Production, just in case there's any regressions found

**Comment by JmillsExpensify** [comment]
Additionally, now that #11063 just hit production, we should remove the hold. Woo!

**Comment by JmillsExpensify** [comment]
cc @trjExpensify Since you're on the main tracking issue.

**Comment by trjExpensify** [comment]
Ah yeah, let me take this over for you @arielgreen. :)

**Comment by trjExpensify** [comment]
@Beamanator what's the plan? Are you taking this or are we going external? I agree it should come off hold now.

**Comment by trjExpensify** [comment]
@Beamanator, this issue will hit 4 weeks old after next week and it doesn't seem like we've made any further progress here yet to get to a resolution before then. What are the next steps?

**Comment by Beamanator** [comment]
True true, I believe next steps are to get this discussion moved to #expensify-open-source to make sure we're aligned on how to proceed

**Comment by Beamanator** [comment]
Posted here to try to get alignment: https://expensify.slack.com/archives/C01GTK53T8Q/p1671547890266159

**Comment by bondydaa** [comment]
Assigning myself since beaman is OOO. caught up on that thread and confirmed with Ioni next steps are:

1. confirm if a png is being converted to a jpg or not?
2. if it is converted to jpg then just make sure it's got the correct file extension. If it stays as a png then make it convert to jpg?
3. ensure all file formats listed already adhere to this https://expensify.slack.com/archives/C01GTK53T8Q/p1671558426183609?thread_ts=1671547931.029679&cid=C01GTK53T8Q 
    1. Converting everything to JPG (in fact, why wouldn’t we also do that for PNG so that everything is in the same file format?) [note from bondy: except for case 1 if it's happening. converting to png to preserve transparency is a nice-to-have for now]
    1. Supporting BMP since it’s a very common file on Windows (and those computers far outnumber Macs)
    1. Not allowing gif animation in avatars, even if we support uploading that file type.


I'll dig into this tomorrow when I'm back online.

**Comment by JmillsExpensify** [comment]
My take is that we should add the `NewFeature` label to this issue. Here's my reasoning: The imagine improvements are like reliable notifications as well as the navigation reboot. They all collective solve "bugs" we've discovered in the app, but they are actually larger architectural plays that are more accurately part of N7. WAQ is simply about solving existing bugs older than 30 days, and this bug was essentially created in the process of adding new capabilities. So sure it's a bug and we should fix it as part of the image improvements project. But I also don't think it's part of WAQ as we result.

**Comment by JmillsExpensify** [comment]
I was chatting with @trjExpensify and he rightly pointed out that we should keep the `bug` label on this issue for clarity. Though in addition to that, I'm going to add `[HOLD WAQ]` in the title, because again, this is bug created by an implementation of new features in progress. So there's no way it should hold up WAQ. Adding the hold will help us stay focused on the issues that block N7 and prevent WAQ from being done.

**Comment by bondydaa** [comment]
okay unassigning myself then and looking for other bugs

**Comment by Beamanator** [comment]
Ok so if this is on hold I shouldn't prioritize it @JmillsExpensify ? Or should I still make progress on it once i'm back (Monday)?

**Comment by trjExpensify** [comment]
Are there other bugs on the image improvements hit list that you're currently working on or is this next up?

**Comment by Beamanator** [comment]
I just submitted a PR for this one today: https://github.com/Expensify/App/issues/13957

So this one is up next!

**Comment by trjExpensify** [comment]
Dope, let's get it!

**Comment by Beamanator** [comment]
Part 1 which will convert all images (except pngs) to jpeg is here: https://github.com/Expensify/Web-Expensify/pull/35994

**Comment by Beamanator** [comment]
Once this is working fine, I'm planning to open this up to contributors to help fix these issues mentioned in the OP:

1. iOS throws errors when uploading BMP
2. iOS converts GIF to JPG in URL of uploaded file
3. Uploading GIF in Android throws console warning, isn't able to upload image

**Comment by trjExpensify** [comment]
Oh, are those three still going to be a problem when part 1 is done? 😕

**Comment by sketchydroide** [comment]
https://github.com/Expensify/App/issues/13910 should be fixed or made not an issue by this GH when it's done, closing the other one.

**Comment by trjExpensify** [comment]
Same [question as here](https://github.com/Expensify/App/issues/13038#issuecomment-1374203405) for @Beamanator, Melv.

**Comment by Beamanator** [comment]
Sorry y'all :D Good questions @trjExpensify - I need to do some testing to see if `bmp` or `gif` have problems when being uploaded in native devices - I don't know exactly _when_ the errors appeared (before or after trying to upload them as avatars). If the errors happened _before_ uploading, I think the issues could still arise

**Comment by Beamanator** [comment]
Ok well I was able to upload a GIF as a chat attachment, but not a profile photo! I can get to the "Edit photo" avatar crop screen, but then clicking "Save" does nothing - this is what we experienced before and mentioned at the top of the issue.

So that will need to be fixed.

> iOS converts GIF to JPG in URL of uploaded file

This is still true, nothing wrong with it :D

> iOS throws errors when uploading BMP

This is also still true, but I thinkkkk we can just not let iOS users upload BMP images instead of spending a lot of time and energy getting a "fix" - I believe it's very unlikely that an iOS user will have BMP images on their phone that they want to upload

**Comment by Beamanator** [comment]
One more thing is I need to test out enabling SVG avatar uploads - i'll work on that today

**Comment by Beamanator** [comment]
Oh it looks like `Image` (React Native component) doesn't support `svg`, this could be a problem

<img width="824" alt="Screenshot 2023-01-11 at 5 18 04 PM" src="https://user-images.githubusercontent.com/3885503/211843946-fc47f372-3c60-4158-8916-7fea9942d83f.png">

**Comment by Beamanator** [comment]
GIF works no problem:

<img width="654" alt="Screenshot 2023-01-11 at 5 21 13 PM" src="https://user-images.githubusercontent.com/3885503/211844652-23ca3932-fac7-4576-91a0-2cd3122126c6.png">

SVG no happy:

`Failed to load resource blob:http://localhost:8080/c19adda9-969b-4303-ab41-bd3e19b7123e (404)`

**Comment by Beamanator** [comment]
@trjExpensify I think I'm going to need someone else's help on this - it kinda looks like there might be a problem storing SVG image data locally (using `URL.createObjectURL`) the way we store everything else, though I couldn't find any evidence of other people hitting the same issue.

I also think it could be nice to wait for Georgia's PR https://github.com/Expensify/App/pull/13216 to be merged before we keep going with SVGs since there's a decent amount of overlap there, getting SVGs to load up in avatar icons

**Comment by JmillsExpensify** [comment]
@Beamanator @trjExpensify We should move the WAQ hold in my opinion in any case.

**Comment by Beamanator** [comment]
Georgia also brought up a good point about svgs:

> Note: The user can still not upload an SVG as an avatar because SVGs must be supported on both old/new dot

**Comment by trjExpensify** [comment]
Question from Friday still stands, Melv.

**Comment by Beamanator** [comment]
Oops sorry for the delay - at least displaying the SVG avatars in OldDot needs to work before we start supporting uploading in NewDot. I'll ping georgia about this

**Comment by grgia** [comment]
For the default avatar PR, since we want to use different default avatars on OldDot vs NewDot, I didn't end up needing to touch OldDot. But yes, to use SVGs for avatars we would need to support it on both platforms

**Comment by trjExpensify** [comment]
I'm a bit confused trying to piece this together:

1. we're saying that NewDot has a [problem with storing SVG avatars locally](https://github.com/Expensify/App/issues/13038#issuecomment-1378982601)
2. Members can't upload SVG avatars because of (1)
3. We're able to use SVG avatars in NewDot for the default avatars despite the limitation in (1)?
4. To do (2) we need to add support for SVG avatars in OldDot as well, that's a concern because of ?? 

We'd probably benefit from writing up a summary of where we're at with these moving parts, adding that to the OP, and starting a discussion on whatever decisions we need to make to go from her.

**Comment by grgia** [comment]
I think it's 
1. I am under the impression that Old Dot does not support SVGs for avatars, and user-uploaded avatars are referenced from the same key in the user's personal details on both Old and NewDot. So if you uploaded an SVG in NewDot (say we store the function), it could work, but then your avatar would be broken when you log back into Old Dot.
2. Yes
3. Yes, the way that we were handling default avatars in New Dot was by hashing logins. So right now, it technically doesn't matter if you have a path to your default avatar saved or not as long as we know the login. Therefore, we were able to store all the default avatar SVGs locally and just grab the correct one whenever we hashing the login. I was following the patterns we already had in place while implementing that PR specifically. For instance, how we handle the SVG workspace avatars.

**Comment by Beamanator** [comment]
Ok yeah so here's a small recap from my side:
1. `gif` avatar uploads don't work on Android devices
2. `bmp` avatar uploads don't work on iOS devices
3. `svg` avatar uploads:
    1. Mobile: Looks like they're not supported in React Native ([notes](https://github.com/Expensify/App/issues/13038#issuecomment-1378927756))
    2. Web:
        - When uploading, we first store the image data in Onyx (using [`URL.createObjectURL()`](https://developer.mozilla.org/en-US/docs/Web/API/URL/createObjectURL)).
        - Somehow this seems to not be working properly, as I keep seeing [this error](https://github.com/Expensify/App/issues/13038#issuecomment-1378974006) when we try to display that immediately after upload
        - Note: Once the SVG uploads, we actually do successfully convert it to a `.jpeg` so actually it should be a problem displaying these in OldDot, so I think that's my bad for leading us down the wrong path there 🙃

**Comment by JmillsExpensify** [comment]
> `gif` How do we feel about disallowing this type of upload for now? 

I think that's fine. These aren't a priority

> `bmp` How do we feel about disallowing this type of upload for now? 

I think that's fine. These aren't a priority

> `svg` Once the SVG uploads, we actually do successfully convert it to a .jpeg so actually it should be a problem displaying these in OldDot

Did you mean **shouldn't** because we converted it to `jpeg`? I think you did but want to confirm.

**Comment by trjExpensify** [comment]
>gif avatar uploads don't work on Android devices
>How do we feel about disallowing this type of upload for now, vs trying to fix the issue in a 3rd party lib?

Isn't it a bit extreme to not allow these file type uploads across all platforms, just because of a problem with the file format on one platform? (Same for `bmp`). Why are we discarding a path like this:  

1. Open it up to a contributor
2. Explain what we want i.e to fix the root cause upstream 
3. Have them submit a PR upstream
4. In the meantime, accept that gif avatars don't work on Android and bmp uploads don't work on iOS until #2.

**Comment by JmillsExpensify** [comment]
That's a fair suggestion, though perhaps we should even align on what doesn't work mean? Does that mean crash? Does that mean the user tries something and it doesn't work with no feedback. I'll admit, I was assuming something worse than just it doesn't work with no feedback.

**Comment by JmillsExpensify** [comment]
Then as for addressing it, to clear, I absolutely agree we should address it. My bigger concern is ensuring that default avatars have nothing in their way, though per my question above, it sounds like this isn't blocking it? Would love a confirmation from @Beamanator.

**Comment by Beamanator** [comment]
@JmillsExpensify I'm 99% sure there's nothing blocking default avatars here, just the uploading of new custom SVG avatars by users

**Comment by Beamanator** [comment]
> Isn't it a bit extreme to not allow these file type uploads across all platforms, just because of a problem with the file format on one platform? (Same for bmp). 

@trjExpensify just to make sure we're on the same page here, for `bmp` I'm suggesting we only disallow `bmp`s from being uploaded on iOS, not on all devices (same for `gif` on Android only).

But yeah if we're ok with `In the meantime, accept that gif avatars don't work on Android and bmp uploads don't work on iOS until (2).`, then I like your plan of opening up an issue to get them fixed, at their own pace 👍

**Comment by JmillsExpensify** [comment]
Awesome that sounds like a plan to me too. Shall I create the separate issue for `bmp` (iOS) and `gif` (Android)?

**Comment by Beamanator** [comment]
Nice, yeah if you have time today that would be great 👍 👍  If not, I can get to it tomorrow or Friday

**Comment by trjExpensify** [comment]
Nice, I think you probably have the most context for the issues to create with the sleuthing to date and what we're after upstream, so probably clearer if you go ahead and create those. 👍

**Comment by Beamanator** [comment]
Makes sense, I was OOO kinda unexpectedly yesterday so I'll try to get to this early next week 👍

**Comment by Beamanator** [comment]
Arg sorry will try to do this tomorrow, it shouldn't take long

**Comment by Beamanator** [comment]
@JmillsExpensify @trjExpensify ok I made this new issue for `bmp` and `gif` avatar images here: https://github.com/Expensify/App/issues/14751

For this issue, I think we can make this specifically be for `.svg` avatar images, update the title & OP, then open externally

**Comment by Beamanator** [comment]
I think this is ready to be worked Externally, so I'll proceed with that plan for now 👍

**Comment by Beamanator** [comment]
Haha yes very true @PankajAS - I'll note in the reproduction steps, even if you add `'svg'` to that list, it doesn't behave as we'd expect

**Comment by Beamanator** [comment]
@PankajAS I updated the OP to include that step

**Comment by PankajAS** [comment]
> @PankajAS I updated the OP to include that step

## Purposal 
Ok @Beamanator for solution SVG need to convert into JPG or PNG image and this it should be perform on backend side. on frontend side we can check if its SVG then it will send to API and in response get png/jpg to show in crop editor.

**Comment by s77rt** [comment]
I think the expected result is to convert svg to png, is that right?

svg is not working on web for many(?) reasons for instance `getImageResolution` returning 0 as height and width.

**Comment by Santhosh-Sellavel** [comment]
@Ollyws Proposal seems to work partially solves the issue showing image immediately after upload

https://user-images.githubusercontent.com/85645967/217301446-24b5e97f-99d4-4a7e-aa71-9decb770aa78.mov

> `const imageName = isSvg ? '': props.imageName;`
Not sure why the image name should be empty.

Check the video, once upload successfully, the color of the avatar changed to dark unable to see it. Different from what we saw while uploading.

[file](https://user-images.githubusercontent.com/85645967/217302056-747286dc-6f45-435e-b03e-3b5b810da492.svg)

**Comment by Santhosh-Sellavel** [comment]
@Beamanator Also, `svg` cannot be selected from Android/iOS Platforms, as we allow only pick from Photos/Gallery.

**Comment by s77rt** [comment]
@Santhosh-Sellavel I think you are expected to test on native. On web we already get to preview the image immediately.
@Beamanator Looks like the server does not support svg images yet (returns a black image instead)

**Comment by Santhosh-Sellavel** [comment]
Platforms reproducible is all. I confirmed issue occurs on all Platform except native As there is no way to upload it. 

It was blank for a while, and got updated to the new image, after applying ollyws proposals I can see immediately but after upload successful I can't see As described in my last comment. 

I'll wait for @Beamanator comments and revist to this tomorrow, thanks!

**Comment by Beamanator** [comment]
@Ollyws it does look like the transparent part of the image is turning black, yeah? I imagine that's why the svg helm (ship steering wheel is all black - the SVG itself is only black lines & a transparent background, right?

**Comment by Beamanator** [comment]
As for the server side, right now I believe we are converting SVG images to jpeg, which is probably why everything is getting a solid background. Why black, though? Not sure. I can try to make some tweaks so SVG images are uploaded as PNGs in the backend

**Comment by Beamanator** [comment]
Also with @Ollyws 's changes in my dev env, I don't see any weird issues with the background turning black once the SVG uploads - you still see the background turning black @Santhosh-Sellavel ?


https://user-images.githubusercontent.com/3885503/217537706-bbb619f4-0494-4b89-b57c-f2be83b9f003.mov


https://user-images.githubusercontent.com/3885503/217538137-10ea72a1-b5cb-460c-a718-6f95672e3c59.mov

**Comment by Beamanator** [comment]
@Ollyws I'm using your latest diff, that's nice that that fixes the black background issue 👍 👍 

I think it's fine if the SVG conversion happens locally 👍 

@Santhosh-Sellavel let me know what you think about @Ollyws 's latest proposal [here](https://github.com/Expensify/App/issues/13038#issuecomment-1421155593), I think it's pretty solid 👍

**Comment by Santhosh-Sellavel** [comment]
Sorry didn't get time to look into this one today, will look into it tomorrow!

**Comment by Beamanator** [comment]
Sounds good @Santhosh-Sellavel 👍

**Comment by Santhosh-Sellavel** [comment]
Looks good to me, and works well.

> const imageName = isSvg ? 'fileName.png' : props.imageName;

@Beamanator Are we fine with this hardcoded file name, do we need to do something about it? Ex: replace extension in name

**Comment by MelvinBot** [comment]
📣 @Ollyws You have been assigned to this job by @Beamanator!
Please apply to this job in Upwork and leave a comment on the Github issue letting us know when we can expect a PR to be ready for review 🧑‍💻
Keep in mind: [Code of Conduct](https://github.com/Expensify/App/blob/main/CODE_OF_CONDUCT.md) | [Contributing](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md) 📖

**Comment by Beamanator** [comment]
> Are we fine with this hardcoded file name, do we need to do something about it? Ex: replace extension in name

@Santhosh-Sellavel - it looks like our backend doesn't care about the file name, it creates its own unique filename when uploading to the server & returns that - It does look a bit weird hardcoding it here but I don't see any problem with it

**Comment by Santhosh-Sellavel** [comment]
Cool then we won't need that name check then we could hardcode a fixed name straight away.

**Comment by Beamanator** [comment]
@Ollyws Aah I see 👍 I think that default file name is only used if somehow the App can't access the uploaded filename. So if we can do the same here that would be great 👍 

And yes please add `svg` to that array of constants 👍  (otherwise we won't be able to select svgs to upload 😅 )

**Comment by MelvinBot** [comment]
`Reviewing` label has been removed, please complete the "BugZero Checklist".

**Comment by MelvinBot** [comment]
The solution for this issue has been :rocket: deployed to production :rocket: in version 1.2.72-1 and is now subject to a 7-day regression period :calendar:. Here is the list of pull requests that resolve this issue:

- https://github.com/Expensify/App/pull/15037


If no regressions arise, payment will be issued on 2023-02-23. :confetti_ball:

After the hold period is over and BZ checklist items are completed, please complete any of the applicable payments for this issue, and check them off once done.
- [ ] External issue reporter
- [ ] Contributor that fixed the issue
- [ ] Contributor+ that helped on the issue and/or PR

As a reminder, here are the bonuses/penalties that should be applied for any External issue:
- Merged PR within 3 business days of assignment - 50% **bonus**
- Merged PR more than 9 business days after assignment - 50% **penalty**

**Comment by trjExpensify** [comment]
Did @adelekennedy pick up a regression that's linked there or is this good to be released for payment?

**Comment by Beamanator** [comment]
Yeah and I just tested with an SVG and it worked fine 🤔 

@kbecciv can you please test uploading an SVG again, like you did in this issue: https://github.com/Expensify/App/issues/15189

And if you are unable to upload, please share the SVG image that failed to upload

**Comment by kbecciv** [comment]
@Beamanator  Please attach a new video and SVG file per your request.


https://user-images.githubusercontent.com/93399543/222512009-f7e775d8-cdf7-40a5-aa09-918dc20d9759.mp4

![svg_logo (3)](https://user-images.githubusercontent.com/93399543/222512056-db54f1d7-6aaf-4217-8f38-6d5a84103dde.svg)

**Comment by Santhosh-Sellavel** [comment]
Yes, it's not a valid SVG. @Beamanator

**Comment by Beamanator** [comment]
Aah interesting, how do y'all know it's actually a PNG image? 😅 Anyway, I think we're good to go here, no regressions @trjExpensify 👍

**Comment by Santhosh-Sellavel** [comment]
@Beamanator 
I just tried opening it in a VSCode but it's not opening. SVG images are XML based and should be able to view/edit in  VSCode (any text editor)

**Comment by trjExpensify** [comment]
Okay great, thanks for confirming! Offers for $1,000 have been sent to you both.

**Comment by trjExpensify** [comment]
Ah shit, that was the weekend. Yes, you're right! I've paid you $500 as a bonus.

**Comment by Beamanator** [comment]
Seems we're only waiting on @Santhosh-Sellavel to accept the Upwork job, then we can pay out & close

**Comment by trjExpensify** [comment]
Yep, correct!
